File: nlox/Lox/Environment.cs
class Environment
{
  private Dictionary<string, object?> values = new Dictionary<string, object?>();

  public Environment() {}

  public Environment(Environment enclosing)
  {
    this.Enclosing = enclosing;
  }

  public Environment? Enclosing { get; init; }

  public void Define(string name, object? value)
  {
    values[name] = value;
  }

  public object? Get(Token name)
  {
    if (values.TryGetValue(name.Lexeme, out var value))
    {
      return value;
    }

    if (Enclosing != null)
    {
      return Enclosing.Get(name);
    }

    throw new RuntimeError(name, $"Undefined variable '{name.Lexeme}'");
  }

  public object? GetAt(int distance, string name)
  {
    var ancestor = Ancestor(distance)!;
    return ancestor.values[name];
  }

  private Environment? Ancestor(int distance)
  {
    var environment = this;
    for (int i = 0; i < distance; i++)
    {
      environment = environment?.Enclosing;
    }
    return environment;
  }

  public void Assign(Token name, object? value)
  {
    if (values.ContainsKey(name.Lexeme))
    {
      values[name.Lexeme] = value;
      return;
    }

    if (Enclosing != null)
    {
      Enclosing.Assign(name, value);
      return;
    }

    throw new RuntimeError(name, $"Undefined variable '{name.Lexeme}'");
  }

  public void AssignAt(int distance, Token name, object? value)
  {
    var ancestor = Ancestor(distance)!;
    ancestor.values[name.Lexeme] = value;
  }
}

File: tslox/src/environment.ts
import RuntimeError from "./runtimeerror";
import { Token } from "./token";

export default class Environment {
  private values = new Map<string, unknown>();

  constructor(public enclosing?: Environment) { }

  define(name: string, value: unknown) {
    this.values.set(name, value);
  }
  
  get(name: Token): unknown {
    if (this.values.has(name.lexeme)) {    
      return this.values.get(name.lexeme);
    }
    if (this.enclosing) {
      return this.enclosing.get(name);
    }
    throw new RuntimeError(name, `Undefined variable '${name.lexeme}'`);
  }

  private ancestor(distance: number) {
    let environment: Environment | undefined = this;
    for (let i = 0; i < distance; i++) {
      environment = environment?.enclosing;
    }
    return environment;
  }

  getAt(distance: number, name: string) {
    const ancestor = this.ancestor(distance)!;
    return ancestor.values.get(name);
  }

  assign(name: Token, value: unknown) {
    if (this.values.has(name.lexeme)) {
      this.values.set(name.lexeme, value);
      return;
    }
    if (this.enclosing) {
      this.enclosing.assign(name, value);
      return;
    }
    throw new RuntimeError(name, `Undefined variable '${name.lexeme}'`);
  }

  assignAt(distance: number, name: Token, value: unknown) {
    const ancestor = this.ancestor(distance)!;
    ancestor.values.set(name.lexeme, value);
  }
}
